### DISCUSSION TITLE
Call a function of a provider inside another provider

### DISCUSSION THREAD

**Discussion by vrman** [rrousselGit/riverpod#1611] [Q&A] [answered]
i have a ChangeNotifierProvider , and i want to call a function of another ChangeNotifierProvider inside it, 
first provider is :
```ruby
class ConfigInitializeController extends ChangeNotifier {
  final AuthUserController authUserController;

  ConfigInitializeController({
    required this.authUserController,
  }) {
    getInitializeConfig();
  }

  var currentState = StateEnum.init;

  Future<void> getInitializeConfig() async {
    if (currentState == StateEnum.loading) {
      return;
    }

    currentState = StateEnum.loading;
    notifyListeners();

    try {
      // This function is on another ChangeNotifierProvider 
      await authUserController.getUserInfo();

      currentState = StateEnum.success;
      notifyListeners();
    } catch (e) {
      currentState = StateEnum.error;
      notifyListeners();
    }
  }
}
```
and the second one , which own getUserInfo function is :
```ruby
class AuthUserController extends ChangeNotifier {
  final AuthController authController;
  final GetUserInfoUseCase getUserInfoUseCase;

  AuthUserController({
    required this.authController,
    required this.getUserInfoUseCase,
  });

  var currentState = StateEnum.init;

  AuthUserInfoEntity? userInfo;

  Future<void> getUserInfo() async {
    if (currentState == StateEnum.loading) {
      return;
    }
    currentState = StateEnum.loading;
    notifyListeners();

    final response = await getUserInfoUseCase(NoParams());

    response.fold((failure) {
      currentState = StateEnum.error;
      notifyListeners();
      throw ('Error getUserInfo');
    }, (response) {
      authController.userInfo = response;

      currentState = StateEnum.success;
      notifyListeners();
    });
  }
}
```

i have this scenario in many places in my app, for example in a page i have a form and i need to update some other providers after submitting that form but i get this error message :
```ruby
Providers are not allowed to modify other providers during their initialization.`
The provider _AutoDisposeNotifierProvider<ConfigInitializeController>#c2001 modified AutoDisposeChangeNotifierProvider<AuthUserController>#11106 while building.
```

i need to know how should i fix this in riverpod

**Comment by rrousselGit** [comment, marked as answer]
As this error tells you, you're not allowed to modify other providers during the initialization of a provider.


In your case you're simply triggering a network request. So it probably should be fine to do:

```diff
    currentState = StateEnum.loading;
    notifyListeners();

    try {
+      await Future.delayed(Duration.zero);


      // This function is on another ChangeNotifierProvider 
      await authUserController.getUserInfo();
```

**Reply by vrman** [reply]
thanks Remi,  the previous error is gone but now problem is its going to infinity loop with this message
```ruby
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: A ConfigInitializeController was used after being disposed.
Once you have called dispose() on a ConfigInitializeController, it can no longer be used.
```

and here is providers definitions:
```ruby
final configInitializeProvider = ChangeNotifierProvider.autoDispose(
  (ref) {
    return ConfigInitializeController(
      authUserController: ref.watch(authUserProvider),
    );
  },
);


final authUserProvider = ChangeNotifierProvider.autoDispose(
  (ref) => AuthUserController(
    authController: ref.watch(authProvider),
    getUserInfoUseCase: ref.watch(getUserInfoUseCaseProvider),
  ),
);
````

**Reply by rrousselGit** [reply]
It's a weird architecture to begin with. You probably want a FutureProvider for those instead of a ChangeNotifier

But you could fix the loop by watching `ref.watch(authprovider.notifier)` & same for the other

**Reply by vrman** [reply]
its fixed , im trying to follow Resocoder TDD architecture, im trying to refactor my controllers from getx to riverpod , in getx it was ok to call and update a controller inside another controller,
do you have any plane to add a future to fix this situation ? and is this ok to use ```await Future.delayed(Duration.zero);``` in same situations ? i have alot of same situation in my other controllers.

**Reply by rrousselGit** [reply]
There's nothing that needs fixing in Riverpod. This error exists on purpose.

The problem is the way you're using it. As I said, you probably would be better off using a `FutureProvider` instead of a `ChangeNotifierProvider` for those.

**Reply by vrman** [reply]
in my getx controller i had multi reactive variables inside one controller , thats why i use ChangeNotifier to make something like
that with riverpod, so then i can have one controller for each page, if i understand you correctly you mean that i need to separate my reactive variables to multi providers instead once (to some FutureProvider and StateNotfier), then each page need multi provider for handling reactive variables. am i correct ?
